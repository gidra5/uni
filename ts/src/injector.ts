import { FileMap } from "codespan-napi";
import type { Precedence } from "./ast";
import type { Position } from "./position";

enum Injectable {
  FileMap = "FileMap",
  RootDir = "RootDir",
  ASTNodeNextId = "ASTNodeNextId",
  ASTNodePrecedenceMap = "ASTNodePrecedenceMap",
  ASTNodePositionMap = "ASTNodePositionMap",
}

type InjectableType = {
  [Injectable.FileMap]: FileMap;
  [Injectable.RootDir]: string;
  [Injectable.ASTNodeNextId]: number;
  [Injectable.ASTNodePrecedenceMap]: Map<string, Precedence>;
  [Injectable.ASTNodePositionMap]: Map<string, Position>;
};

const registry = new Map<string, any>();

const register = <const T extends Injectable>(name: T, value: InjectableType[T]) => {
  registry.set(name, value);
};

const inject = <const T extends Injectable>(name: T): InjectableType[T] => {
  if (!registry.has(name)) {
    throw new Error(`Missing injection entry for: ${name}`);
  }

  return registry.get(name);
};

// Register default injectables
register(Injectable.RootDir, process.cwd());
register(Injectable.FileMap, new FileMap());
register(Injectable.ASTNodeNextId, 0);
register(Injectable.ASTNodePrecedenceMap, new Map());
register(Injectable.ASTNodePositionMap, new Map());

export { register, inject, Injectable };
