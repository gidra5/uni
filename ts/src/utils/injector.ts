import { FileMap } from "codespan-napi";
import type { Precedence } from "../ast";
import type { Position } from "./position";
import { PhysicalType, Type } from "../analysis/types/utils";

enum Injectable {
  FileMap = "FileMap",
  RootDir = "RootDir",
  NextId = "NextId",
  PrecedenceMap = "PrecedenceMap",
  PositionMap = "PositionMap",
  TypeMap = "TypeMap",
  PhysicalTypeMap = "PhysicalTypeMap",
  FreeVariablesMap = "FreeVariablesMap",
  BoundVariablesMap = "BoundVariablesMap",
}

type InjectableType = {
  [Injectable.FileMap]: FileMap;
  [Injectable.RootDir]: string;
  [Injectable.NextId]: number;
  [Injectable.PrecedenceMap]: Map<number, Precedence>;
  [Injectable.PositionMap]: Map<number, Position>;
  [Injectable.TypeMap]: Map<number, Type>;
  [Injectable.PhysicalTypeMap]: Map<number, PhysicalType>;
  [Injectable.FreeVariablesMap]: Map<number, string[]>;
  [Injectable.BoundVariablesMap]: Map<number, string[]>;
};

const registry = new Map<string, any>();

const register = <const T extends Injectable>(name: T, value: InjectableType[T]) => {
  registry.set(name, value);
};

const inject = <const T extends Injectable>(name: T): InjectableType[T] => {
  if (!registry.has(name)) {
    throw new Error(`Missing injection entry for: ${name}`);
  }

  return registry.get(name);
};

// Register default injectables
register(Injectable.RootDir, process.cwd());
register(Injectable.FileMap, new FileMap());
register(Injectable.NextId, 0);
register(Injectable.PrecedenceMap, new Map());
register(Injectable.PositionMap, new Map());
register(Injectable.TypeMap, new Map());

export { register, inject, Injectable };
