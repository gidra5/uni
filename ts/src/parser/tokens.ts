import { SystemError } from "../error.js";
import { type Position } from "../position.js";
import { nextId, setPos } from "../utils/index.js";
import { Parser } from "./utils.js";

export const symbols = [
  "::",
  "<-",
  "<-?",
  "->",
  "--",
  "++",
  "+=",
  "-=",
  "*=",
  "/=",
  "^=",
  "%=",
  "!=",
  "==",
  ">=",
  "<=",
  ":=",
  "|>",
  "===",
  "!==",
  "...",
];

// Sort symbols by length in descending order
symbols.sort((a, b) => b.length - a.length);

const specialStringCharTable = {
  "0": "\0",
  r: "\r",
  n: "\n",
  t: "\t",
};

export const specialStringChars = Object.keys(specialStringCharTable);

type ErrorToken = Extract<Token, { type: "number" | "placeholder" }>;
export type Token = (
  | { type: "error"; cause: SystemError; token: ErrorToken }
  | { type: "identifier"; name: string }
  | { type: "number"; value: number }
  | { type: "placeholder" }
  | { type: "newline" }
  | { type: "string" }
  | { type: "multilineString"; intend: string }
) & { id: number };

export type StringToken = (
  | { type: "error"; cause: SystemError; value: string }
  | { type: "segment"; value: string }
  | { type: "lastSegment"; value: string }
) & { id: number };

const hexLiteralError = Parser.do(function* () {
  const id = nextId();
  const pos: Position = yield Parser.span();
  setPos(id, pos);

  const cause = SystemError.invalidHexLiteral(pos);
  const token = { id, type: "number", value: 0 } satisfies ErrorToken;
  return { id: nextId(), type: "error", cause, token } satisfies Token;
});

const octalLiteralError = Parser.do(function* () {
  const id = nextId();
  const pos: Position = yield Parser.span();
  setPos(id, pos);

  const cause = SystemError.invalidOctalLiteral(pos);
  const token = { id, type: "number", value: 0 } satisfies ErrorToken;
  return { id: nextId(), type: "error", cause, token } satisfies Token;
});

const binaryLiteralError = Parser.do(function* () {
  const id = nextId();
  const pos: Position = yield Parser.span();
  setPos(id, pos);

  const cause = SystemError.invalidBinaryLiteral(pos);
  const token = { id, type: "number", value: 0 } satisfies ErrorToken;
  return { id: nextId(), type: "error", cause, token } satisfies Token;
});

const blockCommentError = Parser.do(function* () {
  const id = nextId();
  const pos: Position = yield Parser.span();
  setPos(id, pos);

  const cause = SystemError.unclosedBlockComment(pos);
  const token = { id, type: "placeholder" } satisfies ErrorToken;
  return { id: nextId(), type: "error", cause, token } satisfies Token;
});

const endOfSourceError = function* () {
  const id = nextId();
  const pos: Position = yield Parser.span();
  setPos(id, pos);

  const cause = SystemError.endOfSource(pos);
  const token = { id, type: "placeholder" } satisfies ErrorToken;
  return { id: nextId(), type: "error", cause, token } satisfies Token;
};

const stringLiteralError = function* (value: string) {
  const id = nextId();
  const pos: Position = yield Parser.span();
  setPos(id, pos);
  const cause = SystemError.unterminatedString(pos);
  return { id, type: "error", cause, value } satisfies StringToken;
};

const number = function* (value: string) {
  const id = nextId();
  setPos(id, yield Parser.span());
  return { id, type: "number", value: Number(value) } satisfies Token;
};

const newline = function* () {
  const id = nextId();
  setPos(id, yield Parser.span());
  return { id, type: "newline" } satisfies Token;
};

const identifier = function* () {
  const id = nextId();
  setPos(id, yield Parser.span());

  const name: string = yield Parser.substring();
  return { id, type: "identifier", name } satisfies Token;
};

const placeholder = function* () {
  const id = nextId();
  setPos(id, yield Parser.span());
  return { id, type: "placeholder" } satisfies Token;
};

const string = function* () {
  const id = nextId();
  setPos(id, yield Parser.span());
  return { id, type: "string" } satisfies Token;
};

const multilineString = function* (intend: string) {
  const id = nextId();
  setPos(id, yield Parser.span());
  return { id, type: "multilineString", intend } satisfies Token;
};

const stringSegment = function* (value: string) {
  const id = nextId();
  setPos(id, yield Parser.span());
  return { id, type: "segment", value } satisfies StringToken;
};

const stringLastSegment = function* (value: string) {
  const id = nextId();
  setPos(id, yield Parser.span());
  return { id, type: "lastSegment", value } satisfies StringToken;
};

const parseHexStyleLiteral = (prefix: string, digitRegexp: RegExp, error: Parser<string, Token>) =>
  Parser.do(function* () {
    let value = prefix;
    while ((yield Parser.checkRegexp(digitRegexp)) || (yield Parser.checkString("_"))) {
      const prev = yield Parser.nextChar();
      if (prev !== "_") value += prev;
    }

    if (value === prefix) return yield error;
    return yield* number(value);
  });

const parseBlockComment = function* self() {
  while (!(yield Parser.string("*/"))) {
    if (yield Parser.string("/*")) yield* self();
    else if (yield Parser.isEnd()) return yield blockCommentError;
    else yield Parser.advance();
  }
  return null;
};

export const parseStringToken = Parser.do<string, StringToken>(function* () {
  yield Parser.rememberIndex();
  let value = "";

  while (true) {
    if (yield Parser.checkString('"')) return yield* stringLastSegment(value);
    if (yield Parser.checkString("\\(")) return yield* stringSegment(value);
    if (yield Parser.isEnd()) return yield* stringLiteralError(value);
    if (yield Parser.checkString("\n")) return yield* stringLiteralError(value);

    if (yield Parser.string("\\")) {
      if (yield Parser.isEnd()) continue;

      const char: keyof typeof specialStringCharTable | null = yield Parser.oneOfStrings(...specialStringChars);
      if (char) {
        value += specialStringCharTable[char];
        continue;
      }
    }

    value += yield Parser.nextChar();
  }
});

export const parseMultilineStringToken = (intend: string) =>
  Parser.do<string, StringToken>(function* () {
    yield Parser.rememberIndex();
    let value = "";

    while (true) {
      if (yield Parser.checkString('"""')) return yield* stringLastSegment(value.trimEnd());
      if (yield Parser.checkString("\\(")) return yield* stringSegment(value);
      if (yield Parser.isEnd()) return yield* stringLiteralError(value.trimEnd());

      if (yield Parser.string(intend)) {
        if (yield Parser.isEnd()) continue;
        value += "\n";
        continue;
      }

      // if intendation is not full, skip until the next character
      if (yield Parser.string("\n")) {
        if (yield Parser.isEnd()) continue;
        while ((yield Parser.checkRegexp(/\s/)) && !(yield Parser.checkString("\n"))) yield Parser.advance();
        value += "\n";
        continue;
      }

      if (yield Parser.string("\\")) {
        if (yield Parser.isEnd()) continue;

        // escape sequence
        type Char = keyof typeof specialStringCharTable | null;
        const char: Char = yield Parser.oneOfStrings(...specialStringChars);
        if (char) {
          value += specialStringCharTable[char];
          continue;
        }
      }

      value += yield Parser.nextChar();
    }
  });

export const parseWhitespace = Parser.do<string, Token | null>(function* () {
  yield Parser.rememberIndex();
  let isNewline = false;

  // skip comments and whitespace
  while (true) {
    if (yield Parser.string("/*")) {
      const result = yield* parseBlockComment();
      if (result) return result;
      continue;
    }

    if (yield Parser.checkRegexp(/\s/)) {
      if (yield Parser.checkString("\n")) isNewline = true;
      yield Parser.advance();
      continue;
    }

    if (yield Parser.string("//")) {
      yield Parser.untilString("\n");
      isNewline = true;
      continue;
    }

    break;
  }

  if (isNewline) return yield* newline();
  if (yield Parser.isEnd()) return yield* endOfSourceError();
  return null;
});

export const parseToken = Parser.do<string, Token>(function* () {
  yield Parser.rememberIndex();

  if (yield Parser.string('"""\n')) {
    let intend = "\n";
    while ((yield Parser.checkRegexp(/\s/)) && !(yield Parser.checkString("\n"))) intend += yield Parser.nextChar();

    return yield* multilineString(intend);
  }

  if (yield Parser.string('"')) return yield* string();

  if (yield Parser.string("0x")) return yield parseHexStyleLiteral("0x", /[0-9a-fA-F]/, hexLiteralError);
  if (yield Parser.string("0o")) return yield parseHexStyleLiteral("0o", /[0-7]/, octalLiteralError);
  if (yield Parser.string("0b")) return yield parseHexStyleLiteral("0b", /[01]/, binaryLiteralError);

  if (yield Parser.checkRegexp(/\d/)) {
    let value = "";

    while (yield Parser.checkRegexp(/\d/)) {
      value += yield Parser.nextChar();
      while (yield Parser.string("_")) {}
    }
    if (yield Parser.checkString(".")) value += yield Parser.nextChar();
    while (yield Parser.checkRegexp(/\d/)) {
      value += yield Parser.nextChar();
      while (yield Parser.string("_")) {}
    }

    return yield* number(value);
  }

  if (/\.\d/.test(yield Parser.peekSubstring(2))) {
    yield Parser.advance();
    let value = "0.";

    while (yield Parser.checkRegexp(/\d/)) {
      value += yield Parser.nextChar();
      while (yield Parser.string("_")) {}
    }

    return yield* number(value);
  }

  if (yield Parser.checkRegexp(/[a-zA-Z_]/)) {
    while (yield Parser.regexp(/\w/)) {}

    const ident = yield Parser.substring();
    if (/^_+$/.test(ident)) return yield* placeholder();
    return yield* identifier();
  }

  for (const symbol of symbols) {
    if (yield Parser.string(symbol)) return yield* identifier();
  }

  yield Parser.nextChar();
  return yield* identifier();
});
