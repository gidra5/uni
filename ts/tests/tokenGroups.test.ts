import { specialStringChars } from "../src/parser/tokens.js";
import {
  parseTokenGroups,
  _parseToken,
  TokenGroup,
  TokenGroupKind,
  parseTokenGroup,
} from "../src/parser/tokenGroups.js";
import { beforeEach, describe, expect } from "vitest";
import { it, fc, test } from "@fast-check/vitest";
import { type Arbitrary } from "fast-check";
import { Iterator } from "iterator-js";
import { eventLoopYield, getPos } from "../src/utils/index.js";
import { Injectable, register } from "../src/injector.js";

beforeEach(() => {
  register(Injectable.NextId, 0);
  register(Injectable.PositionMap, new Map());
});

const anyStringArb = fc.string({ size: "large", unit: "binary" });
const stringInsidesArb = anyStringArb.filter((s) => !s.includes("\n") && !s.includes("\\") && !s.includes('"'));
const charArb = fc.string({ minLength: 1, maxLength: 1 });
const notStringSpecialCharArb = charArb.filter((s) => !specialStringChars.includes(s));
const arrayLenArb = <T>(arb: Arbitrary<T>, len: number) => fc.array(arb, { minLength: len, maxLength: len });

function clearIds({ id, ...token }: any) {
  if (token.type === "group") return { ...token, tokens: token.tokens.map(clearIds) };
  if (token.type === "error") return { ...token, token: clearIds(token.token) };
  return token;
}

const testCase = (input: string) => {
  const tokenGroups = parseTokenGroups(input);
  expect(tokenGroups.map(clearIds)).toMatchSnapshot();
};

describe("group kinds", () => {
  describe("pair groups", () => {
    test("parens", () => testCase("(x)"));
    test("brackets", () => testCase("[x]"));
    test("braces", () => testCase("{x}"));
  });

  describe("for group", () => {
    test("for colon", () => testCase("for x in y: z"));
    test("for arrow", () => testCase("for x in y -> z"));
    test("for braces", () => testCase("for x in y { z }"));

    describe.todo("errors", () => {
      test("parseTokens", () => testCase("for x in y {"));
      test("parseTokens", () => testCase("for x in y }"));
      test("parseTokens", () => testCase("for x { z }"));
      test("parseTokens", () => testCase("for x in y"));
      test("parseTokens", () => testCase("for x : z"));
      test("parseTokens", () => testCase("for x -> z"));
      test("parseTokens", () => testCase("for x }"));
      test("parseTokens", () => testCase("for x {"));
      test("parseTokens", () => testCase("{ for x } }"));
      test("parseTokens", () => testCase("for x"));
    });
  });

  describe("while group", () => {
    test("while colon", () => testCase("while x: z"));
    test("while arrow", () => testCase("while x -> z"));
    test("while braces", () => testCase("while x { z }"));
  });

  describe("inject group", () => {
    test("inject colon", () => testCase("inject x: z"));
    test("inject arrow", () => testCase("inject x -> z"));
    test("inject braces", () => testCase("inject x { z }"));
  });

  describe("mask group", () => {
    test("mask colon", () => testCase("mask x: z"));
    test("mask arrow", () => testCase("mask x -> z"));
    test("mask braces", () => testCase("mask x { z }"));
  });

  describe("without group", () => {
    test("without colon", () => testCase("without x: z"));
    test("without arrow", () => testCase("without x -> z"));
    test("without braces", () => testCase("without x { z }"));
  });

  describe("fn group", () => {
    test("fn colon", () => testCase("fn x: z"));
    test("fn arrow", () => testCase("fn x -> z"));
    test("fn braces", () => testCase("fn x { z }"));
    test.todo("fn arrow braces", () => testCase("fn x -> y { z }"));
  });

  describe("if group", () => {
    test("if colon", () => testCase("if y: z"));
    test("if arrow", () => testCase("if y -> z"));
    test("if braces", () => testCase("if y { z }"));
    test.todo("if colon else", () => testCase("if y: z else x"));
    test.todo("if braces else", () => testCase("if y { z } else x"));
  });

  describe.todo("record group", () => {
    test("record braces", () => testCase("record { a }"));
    test("dict braces", () => testCase("dict { a }"));
  });

  describe("match group", () => {
    test("parseTokens", () => testCase("match x { a -> b }"));
  });
});

test.prop([
  fc
    .array(stringInsidesArb)
    .chain((textSegments) =>
      arrayLenArb(
        anyStringArb.filter((s) => !s.includes("(") && !s.includes(")") && !s.includes('"')),
        Math.max(textSegments.length - 1, 0)
      ).map((joins) => [textSegments, joins])
    )
    .map(([strings, joins]) => {
      const literal = Iterator.zip(strings, [...joins.map((x) => `\\(${x})`), ""])
        .flat()
        .join("");
      const value = Iterator.zip(strings, [...joins, ""]);
      return [literal, value.toArray()] as const;
    }),
])("template strings", async ([literal, interpolated]) => {
  await eventLoopYield();
  const value: TokenGroup[] = interpolated
    .flatMap(([text, interpolated]): TokenGroup[] => [
      { type: "string", value: text } as TokenGroup,
      {
        type: "group",
        kind: TokenGroupKind.Parentheses,
        tokens: parseTokenGroup(")")
          .parse(interpolated + ")", { followSet: [], index: 0 })[1]
          .tokens.tokens.map(clearIds),
      } as TokenGroup,
    ])
    .slice(0, -1);
  if (value.length === 0) value.push({ type: "string", value: "" } as TokenGroup);

  const src = `"${literal}"`;
  const startIndex = 0;
  const expectedIndex = src.length;
  const expectedToken = { type: "group", kind: TokenGroupKind.StringTemplate, tokens: value };
  const [{ index }, token] = _parseToken.parse(src, { index: startIndex, followSet: [] });

  expect(index).toBe(expectedIndex);
  expect(clearIds(token)).toEqual(expectedToken);
});

test("parseTokens", () => testCase('42 "Hello" variable ((expr))'));

it.prop([anyStringArb])("parseTokenGroups never throws", (src) => {
  expect(() => parseTokenGroups(src)).not.toThrow();
});

it.prop([anyStringArb])("parseTokenGroups positions are correctly nested", (src) => {
  const tokenGroup = parseTokenGroups(src);

  check(tokenGroup);

  function check(tg: TokenGroup[], start = 0, end = src.length) {
    tg.forEach(function checkToken(tg: TokenGroup) {
      if (tg.type === "group") {
        if ("kind" in tg) {
          const pos = getPos(tg.id)!;
          check(tg.tokens, pos.start, pos.end);
        } else {
          check(tg.tokens, start, end);
        }
        return;
      }
      if (tg.type === "error") {
        checkToken(tg.token);
        return;
      }

      const pos = getPos(tg.id)!;
      expect(pos.start >= start);
      expect(pos.end <= end);
    });
  }
});
