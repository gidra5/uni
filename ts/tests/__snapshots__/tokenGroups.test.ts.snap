// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`group kinds > fn group > fn arrow 1`] = `
[
  {
    "kind": "function",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "arrow",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > fn group > fn arrow braces 1`] = `
[
  {
    "kind": "function",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "tokens": [
          {
            "name": "y",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "z",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > fn group > fn braces 1`] = `
[
  {
    "kind": "function",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "z",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > fn group > fn colon 1`] = `
[
  {
    "kind": "function",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "colon",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > for group > errors > block missing block end 1`] = `
[
  {
    "kind": "for in",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "tokens": [
          {
            "name": "y",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "cause": [Error: Missing closing token],
        "token": {
          "kind": "braces",
          "tokens": [],
          "type": "group",
        },
        "type": "error",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > for group > errors > missing everything 1`] = `
[
  {
    "kind": "for in",
    "tokens": [
      {
        "cause": [Error: Missing closing token],
        "token": {
          "tokens": [
            {
              "name": "x",
              "type": "identifier",
            },
            {
              "cause": [Error: Unexpected end of source],
              "token": {
                "type": "placeholder",
              },
              "type": "error",
            },
          ],
          "type": "group",
        },
        "type": "error",
      },
      {
        "cause": [Error: missing operand],
        "token": {
          "tokens": [],
          "type": "group",
        },
        "type": "error",
      },
      {
        "cause": [Error: missing operand],
        "token": {
          "tokens": [],
          "type": "group",
        },
        "type": "error",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > for group > for arrow 1`] = `
[
  {
    "kind": "for in",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "tokens": [
          {
            "name": "y",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "arrow",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > for group > for braces 1`] = `
[
  {
    "kind": "for in",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "tokens": [
          {
            "name": "y",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "z",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > for group > for colon 1`] = `
[
  {
    "kind": "for in",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "tokens": [
          {
            "name": "y",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "colon",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > if group > if arrow 1`] = `
[
  {
    "kind": "if",
    "tokens": [
      {
        "tokens": [
          {
            "name": "y",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "arrow",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > if group > if braces 1`] = `
[
  {
    "kind": "if",
    "tokens": [
      {
        "tokens": [
          {
            "name": "y",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "z",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > if group > if colon 1`] = `
[
  {
    "kind": "if",
    "tokens": [
      {
        "tokens": [
          {
            "name": "y",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "colon",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > inject group > inject arrow 1`] = `
[
  {
    "kind": "inject",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "arrow",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > inject group > inject braces 1`] = `
[
  {
    "kind": "inject",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "z",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > inject group > inject colon 1`] = `
[
  {
    "kind": "inject",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "colon",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > mask group > mask arrow 1`] = `
[
  {
    "kind": "mask",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "arrow",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > mask group > mask braces 1`] = `
[
  {
    "kind": "mask",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "z",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > mask group > mask colon 1`] = `
[
  {
    "kind": "mask",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "colon",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > match group > match 1`] = `
[
  {
    "kind": "match",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "a",
            "type": "identifier",
          },
          {
            "name": "->",
            "type": "identifier",
          },
          {
            "name": "b",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > pair groups > braces 1`] = `
[
  {
    "kind": "braces",
    "tokens": [
      {
        "name": "x",
        "type": "identifier",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > pair groups > braces newlines 1`] = `
[
  {
    "kind": "braces",
    "tokens": [
      {
        "name": "x",
        "type": "identifier",
      },
      {
        "type": "newline",
      },
      {
        "name": "y",
        "type": "identifier",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > pair groups > brackets 1`] = `
[
  {
    "kind": "brackets",
    "tokens": [
      {
        "name": "x",
        "type": "identifier",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > pair groups > parens 1`] = `
[
  {
    "kind": "parentheses",
    "tokens": [
      {
        "name": "x",
        "type": "identifier",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > while group > while arrow 1`] = `
[
  {
    "kind": "while",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "arrow",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > while group > while braces 1`] = `
[
  {
    "kind": "while",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "z",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > while group > while colon 1`] = `
[
  {
    "kind": "while",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "colon",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > without group > without arrow 1`] = `
[
  {
    "kind": "without",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "arrow",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`group kinds > without group > without braces 1`] = `
[
  {
    "kind": "without",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "braces",
        "tokens": [
          {
            "name": "z",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;

exports[`group kinds > without group > without colon 1`] = `
[
  {
    "kind": "without",
    "tokens": [
      {
        "tokens": [
          {
            "name": "x",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
      {
        "kind": "colon",
        "tokens": [],
        "type": "group",
      },
    ],
    "type": "group",
  },
  {
    "name": "z",
    "type": "identifier",
  },
]
`;

exports[`parseTokens 1`] = `
[
  {
    "type": "number",
    "value": 42,
  },
  {
    "kind": "string template",
    "tokens": [
      {
        "type": "string",
        "value": "Hello",
      },
    ],
    "type": "group",
  },
  {
    "name": "variable",
    "type": "identifier",
  },
  {
    "kind": "parentheses",
    "tokens": [
      {
        "kind": "parentheses",
        "tokens": [
          {
            "name": "expr",
            "type": "identifier",
          },
        ],
        "type": "group",
      },
    ],
    "type": "group",
  },
]
`;
